contracts/src/token/erc20/extensions/permit.rs

```diff
@@ -0,0 +1,196 @@
+//! Permit Contract.
+//!
+//! Extension of the ERC-20 standard allowing approvals to be made
+//! via signatures, as defined in EIP-2612.
+//!
+//! Adds the `permit` method, which can be used to change an account’s
+//! ERC20 allowance (see [`crate::token::erc20::IErc20::allowance`])
+//! by presenting a message signed by the account.
+//! By not relying on [`crate::token::erc20::IErc20::approve`],
+//! the token holder account doesn’t need to send a transaction,
+//! and thus is not required to hold Ether at all.
+use alloy_primitives::{b256, keccak256, Address, B256, U256};
+use alloy_sol_types::{sol, SolType};
+use stylus_proc::{external, sol_storage, SolidityError};
+use stylus_sdk::{block, prelude::StorageType, storage::TopLevelStorage};
+
+use crate::{
+    token::erc20::IErc20Internal,
+    utils::{
+        cryptography::{ecdsa, eip712::IEIP712},
+        nonces::Nonces,
+    },
+};
+
+const PERMIT_TYPEHASH: B256 =
+    b256!("6e71edae12b1b97f4d1f60370fef10105fa2faae0126114a169c64845d6126c9");
+
+type StructHashTuple = sol! {
+    tuple(bytes32, address, address, uint256, uint256, uint256)
+};
+
+/// Calculates `keccak256` hash value for a struct used in Permit extension.
+///
+/// # Arguments
+///
+/// * `owner` - Account that owns the tokens.
+/// * `spender` - Account that will spend the tokens.
+/// * `value` - The number of tokens being permitted to transfer by `spender`.
+/// * `deadline` - Deadline for the `permit` action.
+pub fn prepare_struct_hash(
+    owner: Address,
+    spender: Address,
+    value: U256,
+    nonce: U256,
+    deadline: U256,
+) -> B256 {
+    keccak256(StructHashTuple::encode_params(&(
+        *PERMIT_TYPEHASH,
+        owner,
+        spender,
+        value,
+        nonce,
+        deadline,
+    )))
+}
+
+sol! {
+    /// Indicates an error related to the fact that
+    /// permit deadline has expired.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC2612ExpiredSignature(uint256 deadline);
+
+    /// Indicates an error related to the issue about mismatched signature.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC2612InvalidSigner(address signer, address owner);
+}
+
+/// A Permit error.
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Indicates an error related to the fact that
+    /// permit deadline has expired.
+    ExpiredSignature(ERC2612ExpiredSignature),
+    /// Indicates an error related to the issue about mismatched signature.
+    InvalidSigner(ERC2612InvalidSigner),
+}
+
+sol_storage! {
+    /// State of a Permit Contract.
+    pub struct Permit<T: IEIP712 + StorageType>{
+        /// Nonces contract.
+        Nonces nonces;
+        /// EIP-712 contract. Must implement [`IEIP712`] trait.
+        T eip712;
```

Comment:
Could you expand on why this needs to be in storage? Why can't we just implement the trait for `Permit` and that's it?

Reply:
Seems we want to let user define his own `NAME` and `VERSION` constants. Those constants are used inside other functions of `IEip712` trait.
The other approach could be: have those fields as strings inside contract storage as `Metadata` extension. And init them through `koba` solidity constructor.

Reply:
Yep, I want to let developers define own `NAME` and `VERSION` while using our `Permit`.